### ISSUE TITLE
Decoding panics with slice range error

### ISSUE THREAD

**Issue by zaynetro** [y-crdt/y-crdt#182]
I have encountered a panic when decoding an update:

```
   6: <core::ops::range::Range<usize> as core::slice::index::SliceIndex<[T]>>::index
             at /rustc/eaadb8947b850a025404082f6297766c4680a42a/library/core/src/slice/index.rs:304:13
   7: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index
             at /rustc/eaadb8947b850a025404082f6297766c4680a42a/library/core/src/slice/index.rs:18:9
   8: yrs::updates::decoder::DecoderV2::read_buf
             at /var/home/roman/.cargo/registry/src/github.com-1ecc6299db9ec823/yrs-0.11.2/src/updates/decoder.rs:268:22
   9: yrs::updates::decoder::DecoderV2::new
             at /var/home/roman/.cargo/registry/src/github.com-1ecc6299db9ec823/yrs-0.11.2/src/updates/decoder.rs:220:29
  10: <yrs::updates::decoder::DecoderV2 as core::convert::From<&[u8]>>::from
             at /var/home/roman/.cargo/registry/src/github.com-1ecc6299db9ec823/yrs-0.11.2/src/updates/decoder.rs:282:9
  11: yrs::updates::decoder::Decode::decode_v2
             at /var/home/roman/.cargo/registry/src/github.com-1ecc6299db9ec823/yrs-0.11.2/src/updates/decoder.rs:21:27
```

Test case:

```
let data = vec![5, 188, 105, 167, 66, 82, 152, 165, 88, 130, 21, 227, 45, 207, 51, 164, 187, 138, 141, 203, 253, 226, 254, 210, 203, 107, 144, 128, 1, 178, 184, 152, 248, 58, 29, 39, 108, 238, 253, 249, 114, 19, 10, 154, 21, 7, 121, 114, 238, 41, 127, 186, 75, 179, 251, 217, 98, 110, 225, 156, 219, 246, 108];
let update = Update::decode_v2(data.as_slice()).unwrap();
```

I have tried Y.js and it also throws `error: Uncaught RangeError: Invalid typed array length: 13500`.

I encode my document with: `doc.encode_state_as_update_v2(&StateVector::default())`

There is a chance that my update data got corrupted or I am using too big client ID. I am still reporting this issue since it would be nice to catch panics on the Rust side.

**Comment by Horusiath** [comment]
This might be a bug on our side. I'll take a look at it.

**Comment by zaynetro** [comment]
OK. As I suspected my update was malformed. Forgot to decrypt the data in one branch :innocent: . 

Anyway, I thing yrs should not panic in such cases.

**Comment by Horusiath** [comment]
Tbh. it feels like panics are the most desired outcome here. It's clearly the case of a bug that was created as a result of either:

1. A developer mistake 
2. Potential input that has been malformed by malicious user action

I guess you could argue that we don't want to panic on eg. server side in case of pt.2, but is this really a case here?

**Comment by zaynetro** [comment]
Yeah, it's when I see a function that returns a `Result` I assume it doesn't panic :) Unfortunately, we don't live in ideal world.

I tried looking at the code in question but couldn't really wrap my head around what happens there. If you feel like it is too hard/inconvenient to fix that function not to panic then just close this issue. It would be nice to document that the method could panic though.

Thank you for your time!

**Comment by Horusiath** [comment]
It may be indeed a proper rationale. I'll try to make it.

**Comment by Horusiath** [comment]
This will be addressed in #187 - however keep in mind that lib0 v2 encoding should be used with caution, depending on the payload characteristics.

**Comment by zaynetro** [comment]
Great work! Thank you!
